src/link/edit.js

```diff
@@ -0,0 +1,122 @@
+import { useEffect, useState } from '@wordpress/element';
+import { RichTextToolbarButton } from '@wordpress/block-editor';
+import {
+	create,
+	insert,
+	useAnchorRef,
+	toggleFormat,
+	applyFormat,
+	removeFormat,
+} from '@wordpress/rich-text';
+import { InlineEditUI } from './inline';
+
+const formatType = 'wikipediapreview/link';
+const formatTitle = 'Wikipedia Preview'; // @todo i18n
+let forceFormVisibleValue;
+
+const Edit = ( {
+	isActive,
+	activeAttributes,
+	contentRef,
+	value,
+	onChange,
+} ) => {
+	const [ isFormVisible, setFormVisible ] = useState( true );
+	const anchorRef = useAnchorRef( {
+		ref: contentRef,
+		value,
+		settings,
+	} );
+
+	// hacky solution to show form after it has been shown before
+	// @todo fix this?
+	useEffect( () => {
+		if ( forceFormVisibleValue < 1 ) forceFormVisibleValue++;
+		else if ( forceFormVisibleValue >= 1 ) setFormVisible( true );
+	}, [ value ] );
+
+	const toggleWP = () => {
+		onChange(
+			toggleFormat( value, {
+				type: formatType,
+			} )
+		);
+		setFormVisible( ! isActive );
+	};
+
+	const insertText = ( selectedValue, title, lang ) => {
+		const toInsert = applyFormat(
+			create( { text: title } ),
+			{
+				type: formatType,
+				attributes: {
+					preview: '',
+					title,
+					lang,
+				},
+			},
+			0,
+			title.length
+		);
+		onChange( insert( value, toInsert ) );
+		forceFormVisibleValue = 0;
+	};
+
+	const updateAttributes = ( selectedValue, title, lang ) => {
+		onChange(
+			applyFormat( selectedValue, {
+				type: formatType,
+				attributes: {
+					preview: '',
+					title,
+					lang,
+				},
+			} )
+		);
+		setFormVisible( false );
+		forceFormVisibleValue = 0;
+	};
+
+	const removeAttributes = () => {
+		onChange( removeFormat( value, formatType ) );
+		setFormVisible( false );
+		forceFormVisibleValue = 0;
+	};
+
+	return (
+		<>
+			<RichTextToolbarButton
+				icon="editor-code"
+				title={ `${ formatTitle } (${ isActive ? 'ON' : 'OFF' })` }
+				isActive={ isActive }
+				onClick={ toggleWP }
+			/>
+			{ isFormVisible && isActive && (
```

Comment:
yes, and `isFormVisible` is the interal state within our format.

```js
24: const [ isFormVisible, setFormVisible ] = useState( true );
```

if you notice line 24, this doesn't make sense, it should be invisible `false` in the beginning, right?
when we have the search/preview UI then we may require more state to control the visibility.

To get more understanding, i suggest trying `false` in this line and then deep into the code, and you will see some tricky part.

In this line 94, the best is `isFormVisible || isActive || anotherVisibleState` to control the `inline` component, we will need it when we have search and preview component UI.

Reply:
Thanks this is really useful, lots to digest. This is what I see when `isFormVisible` is initialized as `false`:
  * when first clicking on  a word: `isActive === false`, `isFormVisible === false`, and no popover is displayed
  * then when clicking on the plugin icon: both `isFormVisible` and `isActive` as `true`, popover shows
  * click add, then `isFormVisible === false`, `isActive === true`, popover is not displayed anymore
  * so far, compared to when `isFormVisible` is initialized as `true`, everything seems to be the same
  * but then if I save the post, view it, then come back to edit it:
    * clicking on the just added preview does not display the popover (because `isFormVisible` is again initialized as `false`)

I think I can see what you mean by eventually needing `anotherVisibleState` to control the visibility of what's rendered